rs/execution_environment/src/metrics.rs

```diff
@@ -268,8 +274,40 @@ impl QueryHandlerMetrics {
                 "The total number of transient errors accumulated \
                         during the query execution",
             ),
+            subnet_query_messages: metrics_registry.histogram_vec(
+                "execution_subnet_query_message_duration_seconds",
+                "Duration of a subnet query message execution, in seconds.",
+                decimal_buckets(-3, 2),
+                &["method_name", "status"],
+            ),
         }
     }
+
+    pub fn observe_subnet_query_message<T>(
+        &self,
+        query_method: QueryMethod,
+        duration: f64,
+        result: &Result<T, UserError>,
+    ) {
+        // Determine the label based on whether the method is a subnet (management canister) query.
+        let method_name_label = &format!(
+            "query_{}{}",
+            if query_method.is_management_canister_query() {
+                "ic00_"
+            } else {
+                ""
+            },
+            query_method
+        );
+        let status_label = match result {
+            Ok(_) => SUCCESS_STATUS_LABEL,
+            Err(user_error) => &format!("{:?}", user_error),
```

Comment:
I'd recommend we use the error code only and not the whole user error (contains the description too) as this can lead to very long labels.